File: bench/match.lua
local function apply(tag, n, val)
    if tag == 1 then return val + n
    elseif tag == 2 then return val - n
    else return val
    end
end

local result = 0
for i = 1, 10000000 do
    result = apply(1, 3, result)
    result = apply(2, 1, result)
    result = apply(3, 0, result)
end
print(result)

File: bench/match.py
class Add:
    def __init__(self, n): self.n = n
class Sub:
    def __init__(self, n): self.n = n
class Nop:
    pass

def apply(op, val):
    if isinstance(op, Add): return val + op.n
    if isinstance(op, Sub): return val - op.n
    return val

a = Add(3)
s = Sub(1)
n = Nop()

result = 0
for i in range(10000000):
    result = apply(a, result)
    result = apply(s, result)
    result = apply(n, result)
print(result)
